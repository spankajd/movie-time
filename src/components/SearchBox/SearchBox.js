import React, { useState, useEffect } from "react";
import { useDispatch } from 'react-redux';

import './SearchBox.scss';
import { setSearchQuery } from "../../store/actions/movieActions";

const SearchBox = ({movies}) => {
    const dispatch = useDispatch();
    const [query, setQuery] = useState("");
    const [filteredSuggestions, setFilteredSuggestions] = useState([]);
    const [showSuggestions, setShowSuggestions] = useState(false);
    const [suggestions , setSuggestions] = useState([]);
    
    useEffect( () => {
        let tempTitle = [];
        movies?.map( movie => {
            tempTitle.push(movie.title);
        })
        setSuggestions(tempTitle);
    },[movies]);


    const handleInputChange = (e) => {
        const userInput = e.target.value;
        setQuery(userInput);
        if (userInput) {
          const filtered = suggestions.filter((item) =>
            item.toLowerCase().includes(userInput.toLowerCase())
          );
          setFilteredSuggestions(filtered);
          setShowSuggestions(true);
        } else {
          setFilteredSuggestions([]);
          setShowSuggestions(false);
        }
        dispatch(setSearchQuery(""));
    };

    const handleSuggestionClick = (suggestion) => {
        setQuery(suggestion);
        setFilteredSuggestions([]);
        setShowSuggestions(false);
        dispatch(setSearchQuery(suggestion));
    };
    const renderSuggestions = () => {
        if (!showSuggestions || filteredSuggestions.length === 0) {
        return null;
    }

    return (
      <ul className="suggestions">
        {filteredSuggestions.map((suggestion, index) => (
          <li key={index} onClick={() => handleSuggestionClick(suggestion)}>
            {suggestion}
          </li>
        ))}
      </ul>
    );
  };

    return <div className="SearchBox">
        <input
        data-testid="SearchBox"
        type="text"
        value={query}
        onChange={handleInputChange}
        placeholder="Search movie..."
        />
        {renderSuggestions()}
    </div>
}

export default SearchBox;