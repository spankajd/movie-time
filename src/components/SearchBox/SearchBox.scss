.SearchBox {
    position: relative;
    flex: 1;
  
  
  input {
    width: 100%;
    padding: 10px;
    font-size: 16px;
    box-sizing: border-box;
    border: none;
    outline: none;
  }
  
  .suggestions {
    position: absolute;
    top: 100%;
    left: 0;
    right: 0;
    background: #FFF;
    border: 1px solid #ccc;
    border-top: none;
    max-height: 150px;
    overflow-y: auto;
    list-style-type: none;
    margin: 0;
    padding: 0;
  }
  
  .suggestions li {
    padding: 10px;
    cursor: pointer;
  }
  
  .suggestions li:hover {
    background: #f0f0f0;
  }


  
  @media (max-width: 768px) {
    width: 100%;
    font-size: 14px;
    input {
        font-size: 14px;
    }
  }
}