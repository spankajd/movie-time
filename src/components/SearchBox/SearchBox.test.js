import { fireEvent, screen } from "@testing-library/react";
import '@testing-library/jest-dom';
import { renderWithProviders } from "../../utils/test-utlis";
import SearchBox from "./SearchBox";

describe("SearchBox Component", () => {

    const dummyMovies = [
        {
            title:"Sample Movie",
            episode_id:1
        },
        {
            title:"Testing Movie",
            episode_id:2
        }
    ]

    it("renders SearchBox", () => {
        renderWithProviders(<SearchBox/>);
        const element = screen.getByTestId(/SearchBox/i);
        expect(element).toBeInTheDocument();
      });
      
    it("matches snapshot", () => {
        const { asFragment } = renderWithProviders(<SearchBox />);
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders SearchBox with handleSuggestionClick triggered", () => {
        renderWithProviders(<SearchBox movies={dummyMovies} />, {
            preloadedState: {
                moviesState: {
                    movies:dummyMovies
                }
            }
        });
        const input = screen.getByTestId(/SearchBox/i);
        fireEvent.change(input,{
            target: {
                value: "Movie"
            }
        });
        const suggestionItem = screen.getByText(/Sample Movie/i);
        fireEvent.click(suggestionItem);
        const element = screen.getByTestId(/SearchBox/i);
        expect(element.value).toBe("Sample Movie");
    });

    it("renders SearchBox with handleInputChange triggered with empty string", () => {
        renderWithProviders(<SearchBox movies={dummyMovies} />, {
            preloadedState: {
                moviesState: {
                    movies:dummyMovies
                }
            }
        });
        const input = screen.getByTestId(/SearchBox/i);
        fireEvent.change(input,{
            target: {
                value: "Movie"
            }
        });

        fireEvent.change(input,{
            target: {
                value: ""
            }
        });
        const element = screen.getByTestId(/SearchBox/i);
        expect(element).toBeInTheDocument();
    });
});