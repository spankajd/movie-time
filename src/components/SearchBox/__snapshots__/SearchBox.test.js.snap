// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBox Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="SearchBox"
  >
    <input
      data-testid="SearchBox"
      placeholder="Search movie..."
      type="text"
      value=""
    />
  </div>
</DocumentFragment>
`;
